spv/sync.go

```diff
@@ -49,6 +49,7 @@ type Syncer struct {
 
 	connectingRemotes map[string]struct{}
 	remotes           map[string]*p2p.RemotePeer
+	remoteAvailable   chan struct{}
```

Comment:
naming this `waitingOnRemote` would be clearer I think.  Remotes might always be available, but this channel will only be non-nil if there are currently waiters currently calling `waitForRemote`.

Reply:
or perhaps that is fully accurate, upon closer inspection it might still be non-nil but closed, as it is only ever set to nil when a peer is removed.

this code is giving me a rough time :) but trying to make sure that we don't end up causing a busy loop in `waitForRemote`

Reply:
Renamed. 

> as it is only ever set to nil when a peer is removed.

It is set to nil when a peer is _connected_. 

So, the way I reason about this code is the following:

1. Everything is done under the `remotesMu` lock
2. The first caller to `waitForRemote` that finds no peers, creates the channel. Everyone else uses the same channel.
3. Once the first peer connects, it closes the channel, sets the field to nil and adds itself to the `remotes` map.

Given everything is under the `remotesMu` lock, a peer entering `waitForRemote()` is _either_ going to find a peer in the `remotes` map or it's going to use or create the `waitingOnRemote` channel.

Inside `waitingOnRemote` the peer captures the channel to use after unlocking `remotesMu`, so even if an arbitrary number of events (connections/disconnections/waits) happen between the unlocking and checking the captured channel `c`, the syncer should still maintain an invariant of there are either `peers` or `waitingOnRemote` will be used.

I guess I could, instead of relying on the first caller to `waitForRemote()` to create the channel, always create it when the last peer disconnects (thus the invariant would be that either `(len(remotes) > 0) != (waitingOnRemote != nil)` - either there are remotes or `waitingOnRemote` is filled).

I could also change this to a `sync.Cond`, but they don't offer a way to wait with a context.